python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,713 @@
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
+    def to(self, type_var: PrimitiveType):
+        raise NotImplementedError()
+
+    def to_byte_buffer(self):
+        raise NotImplementedError()
+
+
+@singledispatch
+def literal(value: Any):
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        from iceberg.expressions.literals import literal
+        >>> literal(123)
+        <IntegerLiteral(123)>v
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {value}")
+
+
+@literal.register
+def _(value: bool):
+    return BooleanLiteral(value)
+
+
+@literal.register
+def _(value: int):
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@literal.register
+def _(value: float):
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@literal.register
+def _(value: str):
+    return StringLiteral(value)
+
+
+@literal.register
+def _(value: uuid.UUID):
+    return UUIDLiteral(value)
+
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
+
+
+@literal.register
+def _(value: bytearray):
+    return BinaryLiteral(value)
+
+
+@literal.register
+def _(value: Decimal):
+    return DecimalLiteral(value)
+
+
+class BaseLiteral(Literal):
+    """Base literal which has a value and can be converted between types"""
+
+    def __init__(self, repr_string: str, value):
+        self._repr_string = repr_string
+        self.value = value
+        self.byte_buffer = None
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __repr__(self):
+        return self._repr_string
+
+    def __str__(self):
+        return str(self.value)
+
+    def to_byte_buffer(self):
+        # TODO need byte buffer conversion
+        raise NotImplementedError()
+
+
+class ComparableLiteral(BaseLiteral):
+    def __init__(self, repr_string: str, value):
+        super(ComparableLiteral, self).__init__(repr_string, value)
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        if self.value is None:
```

Comment:
Literal values must never be `None`, so that we never have to worry about cases where we are comparing `None`. Can you add a `TypeError` or `ValueError` in the constructor? It is important to have this invariant for the rest of the expressions codebase.

Reply:
gotcha, I think this is a good callout. 

In python I think it's a bit hard to enforce such invariant, there's a few way to set the attribute of class in non-standard way. So I did following thing to make it align with expectation
- add `value` with [property annotation](https://docs.python.org/3/library/functions.html#property) so setting it directly like `StringLiteral.value = None` will fail with `AttributeError: can't set attribute`
- add None check with `TypeError` in BaseLiteral constructor `__init__` call so that it will reject None on initialization
  - introduce the inner structure to `_value` as abstraction layer to store the value after initialization, in python this convey the message of private field but does not enforce it. One can still set to to None via `lit.__dict__['_value'] = None` but they are no longer covered by our code
  - `value` property getter is directly returning `_value`, so does `__str__`